### ISSUE TITLE
[fork-join] Can / How assign task to workers

### ISSUE THREAD

**Issue by riskers** [josdejong/workerpool#422]
It's a wonderful lib for workers!

But Can / How it assign task to workers?

```html
<button id="test">test</button>

<script src="https://www.unpkg.com/workerpool@9.0.3/dist/workerpool.js"></script>
<script>
  const pool = workerpool.pool('./worker.js', {
    maxWorkers: 7,
    minWorkers: 3,
  })

  document.getElementById('test').onclick = async () => {
    const res = await pool.exec('log', [123])
    console.log('res',  res)
  }
</script>
```

```js
// worker.js
importScripts('https://www.unpkg.com/workerpool@9.0.3/dist/workerpool.js')

// calacute function
function log(data) {
  console.log(data);
  return data;
}

workerpool.worker({
  log: log
})
```

[SandboxCode ->](https://codesandbox.io/p/sandbox/eloquent-breeze-z6ryy3?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clqf6nbw100063b6p9tkox4px%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clqf6nbw100023b6p5v4htw5d%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clqf6nbw100033b6p552yl9ho%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clqf6nbw100053b6p0lmk0uig%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clqf6nbw100023b6p5v4htw5d%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clqf6nbw100013b6px1zp89ow%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Findex.html%2522%257D%255D%252C%2522id%2522%253A%2522clqf6nbw100023b6p5v4htw5d%2522%252C%2522activeTabId%2522%253A%2522clqf6nbw100013b6px1zp89ow%2522%257D%252C%2522clqf6nbw100053b6p0lmk0uig%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clqf6nbw100043b6px36l432n%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clqf6nbw100053b6p0lmk0uig%2522%252C%2522activeTabId%2522%253A%2522clqf6nbw100043b6px36l432n%2522%257D%252C%2522clqf6nbw100033b6p552yl9ho%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clqf6nbw100033b6p552yl9ho%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

I see that `workerpool.pool` is executed and 3 workers are created:
<img width="482" alt="image" src="https://github.com/josdejong/workerpool/assets/5653652/9746c27d-545e-41aa-8856-31cd5fcd8566">

There is always only one worker executing when I clicked button:
<img width="298" alt="image" src="https://github.com/josdejong/workerpool/assets/5653652/ee23ed06-0240-44b0-a3a4-9205862370ce">
<img width="377" alt="image" src="https://github.com/josdejong/workerpool/assets/5653652/7ae07ea7-6234-4bce-a562-7782d4da614e">

Can each worker perform the task equally?

---

Because now I have an application scenario:

```js
const pool = workerpool.pool({
  maxWorkers: 7,
  minWorkers: 3,
})

const tasks = [..] // many data
const promiseList = []
for (let i = 0; i < tasks.length; i++) {
  const exec = pool.exec('log', [tasks[i]])
  promiseList.push(exec)
}

const res = await Promise.all(promiseList)
```

But only one worker is executing the task. Am I using it wrong?

I thought it would be possible to execute multiple tasks in multiple workers at the same time and then aggregate the results, like [fork join](https://en.wikipedia.org/wiki/Fork%E2%80%93join_model).

**Comment by josdejong** [comment]
> I thought it would be possible to execute multiple tasks in multiple workers at the same time and then aggregate the results, like [fork join](https://en.wikipedia.org/wiki/Fork%E2%80%93join_model).

The core feature of `workerpool` is that you can execute multiple tasks in parallel, a nice example to try this out is [examples/browser/dedicatedWorker.html](https://github.com/josdejong/workerpool/blob/master/examples/browser/dedicatedWorker.html). If you have a big task that you can split in smaller parts, you can indeed execute the subtasks in parallel and in the end aggregate the results like in your example, you last example with the `await Promise.all` should indeed work.

**Comment by riskers** [comment]
why is there always only one worker execute?

**Comment by josdejong** [comment]
That is not the case.

**Comment by riskers** [comment]
![1](https://github.com/josdejong/workerpool/assets/5653652/eb52a95d-a814-48d6-88ab-d46a45a364ad)


[codesandbox](https://z6ryy3.csb.app/) There is only a second worker executing

**Comment by josdejong** [comment]
`workerpool` only creates more workers when needed:when a new task is added whilst all created workers are busy (and `maxWorkers` isn't reached). The tasks in your example take about no time, so I guess that case does not happen in your example. To see more workers being created, change your tasks to something that takes say 10 seconds (for example return a Promise that resolves after 10 seconds using a `setTimeout`).

**Comment by riskers** [comment]
Oh, I see.

Thank you for your patience 😄

**Comment by riskers** [comment]
Here is [demo](https://codesandbox.io/p/sandbox/eloquent-breeze-z6ryy3?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clqf6nbw100063b6p9tkox4px%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clqf6nbw100023b6p5v4htw5d%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clqf6nbw100033b6p552yl9ho%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clqf6nbw100053b6p0lmk0uig%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clqf6nbw100023b6p5v4htw5d%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clqf6nbw100013b6px1zp89ow%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Findex.html%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clqf6nbw100023b6p5v4htw5d%2522%252C%2522activeTabId%2522%253A%2522clqf6nbw100013b6px1zp89ow%2522%257D%252C%2522clqf6nbw100053b6p0lmk0uig%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clqf6nbw100043b6px36l432n%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clqf6nbw100053b6p0lmk0uig%2522%252C%2522activeTabId%2522%253A%2522clqf6nbw100043b6px36l432n%2522%257D%252C%2522clqf6nbw100033b6p552yl9ho%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clqf6nbw100033b6p552yl9ho%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D).

If anyone meet same question.
